src/compiler/checker.ts

```diff
@@ -8232,9 +8240,9 @@ namespace ts {
                             // Props is of type 'any' or unknown
                             return links.resolvedJsxType = attributesType;
                         }
-                        else if (!(attributesType.flags & TypeFlags.ObjectType)) {
-                            // Props is not an object type
-                            error(node.tagName, Diagnostics.JSX_element_attributes_type_0_must_be_an_object_type, typeToString(attributesType));
+                        else if (attributesType.flags & TypeFlags.Union) {

```

Comment:
I don't think this actually prevents it from being a union type though. It could be a type parameter, which is later instantiated to a union type.